File: example/python/sensor_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
MagicDog SDK Python 使用示例

这个文件展示了如何使用 MagicDog SDK 的 Python 绑定来控制机器人。
"""

import sys
import time

# 添加构建目录到 Python 路径，插入到最前面
sys.path.insert(0, '../../build')

try:
    import magicdog_python as magicdog
    from magicdog_python import TtsCommand, TtsPriority, TtsMode, GetSpeechConfig, ErrorCode
    print("成功导入 MagicDog Python 模块！")
    # 打印导入的 magicdog_python 的路径
    print(f"导入的 magicdog_python 的路径: {sys.path}")
except ImportError as e:
    print(f"导入失败: {e}")
    print("请先运行 build_python.sh 构建 Python 绑定")
    sys.exit(1)

def main():
    """主函数"""
    print("MagicDog SDK Audio Python 示例程序")
    
    local_ip = "192.168.55.10"
    robot = magicdog.MagicRobot()
    if not robot.initialize(local_ip):
        print("机器人初始化失败")
        return

    robot.set_timeout(5000)
    if not robot.connect():
        print("机器人连接失败")
        robot.shutdown()
        return
    
    print("机器人连接成功")
    
    sensor_controller = robot.get_sensor_controller()
    # sensor_controller.subscribe_tof(lambda tof: print(f"TOF: {len(tof.data)}"))
    sensor_controller.subscribe_ultra(lambda ultra: print(f"Ultra: {len(ultra.data)}"))
    sensor_controller.subscribe_lidar(lambda lidar: print(f"Lidar: {len(lidar.ranges)}"))
    # sensor_controller.subscribe_imu(lambda imu: print(f"IMU: {imu.orientation}"))
    sensor_controller.subscribe_rgbd_color_camera_info(lambda camera_info: print(f"RGBD Color Camera Info: {camera_info.K}"))
    sensor_controller.subscribe_rgbd_depth_image(lambda depth_image: print(f"RGBD Depth Image: {len(depth_image.data)}"))
    sensor_controller.subscribe_rgbd_color_image(lambda color_image: print(f"RGBD Color Image: {len(color_image.data)}"))
    sensor_controller.subscribe_rgb_depth_camera_info(lambda camera_info: print(f"RGB Depth Camera Info: {camera_info.K}"))
    sensor_controller.subscribe_left_binocular_high_img(lambda img: print(f"Left Binocular High Image: {len(img.data)}"))
    sensor_controller.subscribe_left_binocular_low_img(lambda img: print(f"Left Binocular Low Image: {len(img.data)}"))
    sensor_controller.subscribe_right_binocular_low_img(lambda img: print(f"Right Binocular Low Image: {len(img.data)}"))

    status = sensor_controller.open_channel_switch()
    if status.code != ErrorCode.OK:
        print(f"打开通道失败: {status.message}")
        robot.shutdown()
        return

    status = sensor_controller.open_lidar()
    if status.code != ErrorCode.OK:
        print(f"打开激光雷达失败: {status.message}")
        robot.shutdown()
        return
    
    status = sensor_controller.open_rgbd_camera()
    if status.code != ErrorCode.OK:
        print(f"打开RGBD相机失败: {status.message}")
        robot.shutdown()
        return
    
    status = sensor_controller.open_binocular_camera()
    if status.code != ErrorCode.OK:
        print(f"打开立体相机失败: {status.message}")
        robot.shutdown()
        return

    time.sleep(10)

    status = sensor_controller.close_lidar()
    if status.code != ErrorCode.OK:
        print(f"关闭激光雷达失败: {status.message}")
        robot.shutdown()
        return

    status = sensor_controller.close_rgbd_camera()
    if status.code != ErrorCode.OK:
        print(f"关闭RGBD相机失败: {status.message}")
        robot.shutdown()
        return

    status = sensor_controller.close_binocular_camera()
    if status.code != ErrorCode.OK:
        print(f"关闭立体相机失败: {status.message}")
        robot.shutdown()
        return

    time.sleep(10)

    status = sensor_controller.close_channel_switch()
    if status.code != ErrorCode.OK:
        print(f"关闭通道失败: {status.message}")
        robot.shutdown()
        return

    # 避免 lcm 中有缓冲数据未处理完
    time.sleep(10)

    robot.disconnect()
    robot.shutdown()
    print("\n示例程序执行完成！")

if __name__ == "__main__":
    main() 

File: example/cpp/sensor_example/sensor_example.cpp
#include "magic_robot.h"
#include "magic_type.h"

#include <unistd.h>
#include <csignal>

#include <iostream>
#include <memory>

using namespace magic::dog;

magic::dog::MagicRobot robot;

void signalHandler(int signum) {
  std::cout << "Interrupt signal (" << signum << ") received.\n";

  robot.Shutdown();
  // 退出进程
  exit(signum);
}

int main() {
  // 绑定 SIGINT（Ctrl+C）
  signal(SIGINT, signalHandler);

  std::string local_ip = "192.168.55.10";
  // 配置本机网线直连机器的IP地址，并进行SDK初始化
  if (!robot.Initialize(local_ip)) {
    std::cerr << "robot sdk initialize failed." << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 设置rpc超时时间为5s
  robot.SetTimeout(5000);

  // 连接机器人
  auto status = robot.Connect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "connect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  auto& controller = robot.GetSensorController();

  status = controller.OpenChannelSwith();
  if (status.code != ErrorCode::OK) {
    std::cerr << "open channel failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  // controller.SubscribeTof([](const std::shared_ptr<Tof>) {
  //   static unsigned int count = 0;
  //   if (count++ % 10 == 0) {
  //     std::cout << "receive tof." << std::endl;
  //   }
  // });
  // controller.SubscribeUltra([](const std::shared_ptr<Ultra>) {
  //   static unsigned int count = 0;
  //   if (count++ % 10 == 0) {
  //     std::cout << "receive ultra." << std::endl;
  //   }
  // });
  // controller.SubscribeHeadTouch([](const std::shared_ptr<HeadTouch>) {
  //   std::cout << "receive head touch." << std::endl;
  // });

  // 打开lidar
  // status = controller.OpenLidar();
  // if (status.code != ErrorCode::OK) {
  //   std::cerr << "open lidar failed"
  //             << ", code: " << status.code
  //             << ", message: " << status.message << std::endl;
  //   robot.Shutdown();
  //   return -1;
  // }

  // controller.SubscribeImu([](const std::shared_ptr<Imu> msg) {
  //   static unsigned int count = 0;
  //   if (count++ % 10000 == 0) {
  //     std::cout << "receive imu." << std::endl;
  //   }
  // });

  // controller.SubscribeLidar([] (const std::shared_ptr<LaserScan>) {
  //   std::cout << "receive lidar point cloud." << std::endl;
  // });

  // controller.SubscribeRgbdColorCameraInfo([](const std::shared_ptr<CameraInfo> msg) {
  //   std::cout << "receive rgbd color camera info." << std::endl;
  // });
  // controller.SubscribeRgbdDepthImage([](const std::shared_ptr<Image> msg) {
  //   std::cout << "receive rgbd depth image." << std::endl;
  // });
  // controller.SubscribeRgbdColorImage([](const std::shared_ptr<Image> msg) {
  //   std::cout << "receive rgbd color image." << std::endl;
  // });
  // controller.SubscribeRgbDepthCameraInfo([](const std::shared_ptr<CameraInfo> msg) {
  //   std::cout << "receive rgb depth camera info." << std::endl;
  // });

  // status = controller.OpenRgbdCamera();
  // if (status.code != ErrorCode::OK) {
  //   std::cerr << "open rgbd camera failed"
  //             << ", code: " << status.code
  //             << ", message: " << status.message << std::endl;
  //   robot.Shutdown();
  //   return -1;
  // }

  controller.SubscribeLeftBinocularHighImg([](const std::shared_ptr<CompressedImage> msg) {
    std::cout << "receive left binocular high img." << std::endl;
  });
  controller.SubscribeLeftBinocularLowImg([](const std::shared_ptr<CompressedImage> msg) {
    std::cout << "receive left binocular low img." << std::endl;
  });
  controller.SubscribeRightBinocularLowImg([](const std::shared_ptr<CompressedImage> msg) {
    std::cout << "receive right binocular low img." << std::endl;
  });
  controller.SubscribeDepthImage([](const std::shared_ptr<Image> msg) {
    std::cout << "receive depth image." << std::endl;
  });

  status = controller.OpenBinocularCamera();
  if (status.code != ErrorCode::OK) {
    std::cerr << "open binocular camera failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  usleep(50000000);

  // 关闭lidar
  // status = controller.CloseLidar();
  // if (status.code != ErrorCode::OK) {
  //   std::cerr << "close lidar failed"
  //             << ", code: " << status.code
  //             << ", message: " << status.message << std::endl;
  //   robot.Shutdown();
  //   return -1;
  // }

  // 关闭RGBD相机
  // status = controller.CloseRgbdCamera();
  // if (status.code != ErrorCode::OK) {
  //   std::cerr << "close rgbd camera failed"
  //             << ", code: " << status.code
  //             << ", message: " << status.message << std::endl;
  //   robot.Shutdown();
  //   return -1;
  // }

  // 关闭双目相机
  status = controller.CloseBinocularCamera();
  if (status.code != ErrorCode::OK) {
    std::cerr << "close binocular camera failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
  }

  status = controller.CloseChannelSwith();
  if (status.code != ErrorCode::OK) {
    std::cerr << "close channel failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  // 断开与机器人的链接
  status = robot.Disconnect();
  if (status.code != ErrorCode::OK) {
    std::cerr << "disconnect robot failed"
              << ", code: " << status.code
              << ", message: " << status.message << std::endl;
    robot.Shutdown();
    return -1;
  }

  robot.Shutdown();

  return 0;
}